### ISSUE TITLE
Genie classes can implement an interface without needing to inherit from Object

### ISSUE THREAD

**Issue by astavale** [darkoverlordofdata/shmupwarz-vala#2]
Instantiating an object that inherits from GLib's Object is a bit slower than just instantiating the object. So if you are creating a large number of objects it is more efficient not to inherit from Object, unless you want the features that brings (e.g. signals, advanced property features such as the notify signal). If you are simply inheriting from Object to get Genie to implement the interface, then I think this is a bug in the parser. See https://bugzilla.gnome.org/show_bug.cgi?id=776836

You can implement an interface without inheriting by using the inheritance syntax for the interface, e.g. 

`class PlayerComponent : IComponent`

This gives you the benefit of polymorphism, without needing the full features of GObject. A nice diagram of the differences is here: https://wiki.gnome.org/Projects/Vala/FAQ#What_does_.5BSimpleType.5D_and_.5BCompact.5D_in_bindings_mean.3F

**Comment by darkoverlordofdata** [comment]
Interfaces shouldn't have extra baggage, thanks for pointing that out. This is the prototypal version, and I'm not currently making chanes to it, so I'll leave this open for now - I'm making a shared library using this as reference,  I'll make sure my new interfaces are clean.

Gnome has conficting documentation about which features require GObject and which don't. The Projects Genie page says 'to implement an interface it has to be a sub-type of a GLib Object.' but that is not the case. I found that I could create an alternate Object, and use that in place of GObject, and my first version of the game leveraged this, so I had no dependencies on GObject. 

Then I made a change and needed GObject for some feature, so I just replaced my dummy base class with GObject, and it worked great - I didn't see any degradation. I had initially removed GObject dependencies due to my own performance concerns, so if it had no impact, perhaps I was guilty of premature optimization.

**Comment by astavale** [comment]
> The Projects Genie page says 'to implement an interface it has to be a sub-type of a GLib Object.' but that is not the case.

Thanks for pointing that out. I've improved that a bit by removing that statement from the Tutorial.

**Comment by darkoverlordofdata** [comment]
Ah, the defect was in my generated boilerplate. When I made the change to swap out GObject, I neglected to update the template. So the last time I regenerated the components, it became broken. I included a corrected local template in the project, so I don't do that again.

So, I tried just using this

class BoundsComponent : IComponent 

But I get link errors, unsatisfied dependencies on DarkMatter (my dummy object):


public class DarkMatter  {}
namespace Bosco.ECS 
    public interface IComponent: DarkMatter  {}
}


Then if I remove the dummy DarkMatter object from IComponent, I get another issue:

error: missing class prerequisite for interface `Bosco.ECS.IComponent', add GLib.Object to interface declaration if unsure
        _queue : array of IComponent


So I'm using
class BoundsComponent : DarkMatter implements IComponent
